_posts/2017-08-23-auto-healing-foreman-proxy.md

```diff
@@ -0,0 +1,109 @@
+            "sudo yum install -y qpid-dispatch-router",
+            "sudo yum install -y mongodb-server",
+            "sudo yum install -y pulp-rpm-plugins",
+            "sudo yum install -y pulp-server",
+            "sudo yum install -y pulp-docker-plugins",
+            "sudo yum install -y pulp-selinux",
+            "sudo yum install -y python-pulp-streamer",
+            "sudo yum install -y python-gofer-qpid",
+            "sudo yum install -y ca-certificates",
+            "sudo yum install -y mailcap",
+            "sudo yum install -y python-qpid",
+            "sudo yum install -y python-qpid-qmf",
+            "sudo yum install -y qpid-cpp-client",
+            "sudo yum install -y python-crane",
+            "sudo yum install -y puppet-server",
+            "sudo yum install -y foreman-proxy",
+            "sudo yum install -y qpid-tools",
+            "sudo yum install -y rubygem-apipie-bindings",
+            "sudo yum install -y qpid-cpp-server-linearstore",
+            "sudo yum install -y httpd",
+            "sudo yum install -y mod_passenger",
+            "sudo yum install -y mod_wsgi",
+            "sudo yum install -y mod_xsendfile",
+            "sudo yum install -y mod_ssl",
+            "sudo yum install -y pulp-puppet-plugins",
+            "sudo yum install -y mongodb",
+            "sudo yum install -y puppet",
+            "sudo yum install -y katello-client-bootstrap",
+            "sudo yum install -y qpid-cpp-server",
+            "sudo yum install -y rubygem-json",
+            "sudo yum install -y rubygem-smart_proxy_pulp",
+            "sudo systemctl mask sshd; sudo systemctl disable sshd; sudo systemctl stop sshd"
+        ]
+    }
+],
+
+<SNIP>
+```
+
+We couldn't just do a single long yum command, which would be more efficient, because `yum install -y httpd fake-package` would give a 0 exit code, resulting in a silent failure.
+
+### Terraform, the juicy part
+
+The automated rebuild process uses a randomly assigned hostname, we needed to setup an Elastic Load Balancer attached to an auto scaling group of 1 so we can rebuild without registering to the Katello Server as a new Smart Proxy. Here's how the whole process worked end-to-end:
+
+1. A security group allowing ports required for the Smart Proxy Content.
+1. Upload certs.tar to S3 for our capsules.
+1. An IAM profile to allow the instance to pull down the certs.tar from S3
+1. An Elastic Load Balancer with listeners for all the ports Smart Proxy uses, this ensures Katello & Clients can use the same hostname to communicate with it.
+1. Instance using the above "foreman-proxy" AMI with user data that:
+   * Sets the hostname to that of the ELB
+   * Pulls down the certs.tar from S3
+   * Runs foreman-installer via an [Ansible role](https://galaxy.ansible.com/sean797/foreman_installer)
+   * Hits https://katello.example.com/katello/api/capsules/:id/content/sync to kick off a Content Sync
+   * Reverts hostname change back
```

Comment:
Why is that required?

Reply:
It isn't related to Katello but other reasons like logging software also running on the same machine. Maybe I should remove it, as its not really relevant?

Reply:
Maybe "Reverts hostname change back, to meet expectations of the internal tooling, CMDB, etc"?